diff --git a/pisoni.gemspec b/pisoni.gemspec
index 4f62346..fc89442 100644
--- a/pisoni.gemspec
+++ b/pisoni.gemspec
@@ -19,7 +19,8 @@ Gem::Specification.new do |s|
   s.license     = 'Apache-2.0'

   # faraday v2.9.0 removes support for Ruby 2.7, see https://github.com/lostisland/faraday/releases/tag/v2.9.0
-  s.add_runtime_dependency 'faraday', '~> 2.0', '<= 2.9'
+  # Updated to allow Faraday 2.14.3+ which fixes nested query string vulnerability
+  s.add_runtime_dependency 'faraday', '~> 2.0', '>= 2.14.3'
   s.add_runtime_dependency 'json', '~> 2.7', '>= 2.7.1'
   s.add_runtime_dependency 'injectedlogger', '0.0.13'
   s.add_runtime_dependency 'faraday-net_http_persistent', '~> 2.1'
